### DISCUSSION TITLE
For info: Collection types: some benchmarking

### DISCUSSION THREAD

**Discussion by FullValueRider** [twinbasic/twinbasic#888] [Show and tell]
I was looking for a collection type that would allow me to get the keys as well as the items.  The scripting.dictionary is well known but requires both a key and a value so is not a direct drop in replacement for a Collection.  Hunting around a bit I found the cHashD class on VB forums (https://www.vbforums.com/showthread.php?834515-Simple-and-fast-lightweight-HashList-Class-(no-APIs)), and also an implementation of a Treap in VB that compiles in twinbasic without editing (apologies I can't find the url but the source code is attached)

[treapsrc.zip](https://github.com/WaynePhillipsEA/twinbasic/files/8804603/treapsrc.zip)

**Updated to add Olaf's cCollection from RC6**
Some simple benchmarks for comparison.
```
Executing 'HashD.Main.Test'...
Count of Test-Entries: 50000 
Collection-Add:             67.8672sec
cHashD-Add:                 0.0352sec
Dictionary-Add:             0.0469sec
cTreap-Add:                 0.6172sec
cCollection-Add:            0.0391sec

Collection-ItemByKey:       67.8672sec
cHashD-ItemByKey:           0.0352sec
Dictionary-ItemByKey:       0.0352sec
cTreap-ItemByKey:           0.4180sec
ccollection-ItemByKey:      0.0273sec

Indexed access is not compared (it would be much faster per HashList)
Distribution over a HashTable with 16384 slots:
Used-HashSlots: 16317 
Min-Entries: 1 
Max-Entries: 7 
(time taken: 137.0242806s)
```

Based on these timings I'm currently using a Wrapper for the scripting.dictionary which allows the key to be optional, and if not provided uses the dictionary count as the key.

cHashD bears more investigation but the code does need tidying up.

**Comment by wqweto** [comment]
Btw, you can format time to milliseconds i.e. `Format$(Elapsed, "0.000")` for less scientific notations in the results.

I'm sure Olaf (the author of `cHashD`) has even better and battle tested `Collection` impl in his RC5/RC6.

**Reply by FullValueRider** [reply]
Cheers for the pointer.  I've updated the original post with timings and formatting.  Unfortunately for Olaf I prefer open source code so whilst I'm using a scripting.dictionary wrapper at the moment, I'll very likely move to the cHashD object once I've got my head around the code.
